src/linesensor.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include <vector>
+#include <atomic>
+
+#include <driver/spi_master.h>
+
+namespace mcp3008 {
+
+/**
+ * \brief The LineSensor device - one MCP3008 chip.
+ *        It is not thread-safe, you have to make sure the methods are called
+ *        from one thread at a time only.
+ *        The @{install} method has to be called before you can use any other methods.
+ */
+class LineSensor {
+public:
+    static constexpr int CHANNELS = 8; //!< Amount of channels on the chip
+    static constexpr uint16_t MAX_VAL = 1023; //!< Maximum value returned by from the chip (10bits).
+
+    /**
+     * \brief The LineSensor SPI configuration.
+     */
+    struct Config {
+        Config(int freq = 1350000, spi_host_device_t spi_dev = HSPI_HOST,
+            gpio_num_t pin_cs = GPIO_NUM_25, gpio_num_t pin_mosi = GPIO_NUM_33,
+            gpio_num_t pin_miso = GPIO_NUM_32, gpio_num_t pin_sck = GPIO_NUM_26) {
+            this->freq = freq;
+            this->spi_dev = spi_dev;
+
+            this->pin_cs = pin_cs;
+            this->pin_mosi = pin_mosi;
+            this->pin_miso = pin_miso;
+            this->pin_sck = pin_sck;
+        }
+
+        int freq; //!< SPI communication frequency
+        spi_host_device_t spi_dev; //!< Which ESP32 SPI device to use.
+
+        gpio_num_t pin_cs;
+        gpio_num_t pin_mosi;
+        gpio_num_t pin_miso;
+        gpio_num_t pin_sck;
+    };
+
+    LineSensor();
+    virtual ~LineSensor(); //!< The @{uninstall} method is called from the destructor.
+
+    /**
+     * \brief Initialize the SPI bus. Must be called before any other methods,
+     *        otherwise they will return ESP_FAIL.
+     *
+     * \param cfg the SPI bus configuration.
+     * \return ESP_OK or any error code encountered during the inialization.
+     *         Will return ESP_FAIL if called when already installed.
+     */
+    esp_err_t install(const Config& cfg = Config());
+
+    /**
+     * \brief Free the SPI bus. Must be called last, or other methods return ESP_FAIL.
+     *
+     * \return ESP_OK or any error code encountered during the freeing.
+     *         Will return ESP_FAIL if called when not installed.
+     */
+    esp_err_t uninstall();
+
+    /**
+     * \brief Read values from the chip. Returns values in range <0; @{MAX_VAL}>.
+     *
+     * \param results the results will be APPENDED to this vector.
+     *        It will be unchanged unless the ESP_OK result is returned (except possibly its capacity).
+     *        Between 0 and @{CHANNELS} values are appended, depending on the @{channels_mask} param.
+     * \param differential return differential readings, as specified in the MCP3008 datasheet.
```

Comment:
I don't think that is this parameter good in the LineSensor class. It doesn't make sense to use it on a line sensor board.